src/java.base/share/classes/jdk/internal/classfile/impl/verifier/ParserVerifier.java

```diff
@@ -0,0 +1,373 @@
+            }
+            if (m.methodName().equalsString(CLASS_INIT_NAME)
+                    && !m.flags().has(AccessFlag.STATIC)) {
+                errors.add(new VerifyError("Method <clinit> is not static in %s".formatted(toString(classModel))));
+            }
+            if (classModel.flags().has(AccessFlag.INTERFACE)
+                    && m.methodName().equalsString(INIT_NAME)) {
+                errors.add(new VerifyError("Interface cannot have a method named <init> in %s".formatted(toString(classModel))));
+            }
+            verifyMethodName(m.methodName().stringValue());
+        } catch (VerifyError ve) {
+            errors.add(ve);
+        }
+    }
+
+    private void verifyAttributes(ClassFileElement cfe, List<VerifyError> errors) {
+        if (cfe instanceof AttributedElement ae) {
+            var attrNames = new HashSet<String>();
+            for (var a : ae.attributes()) {
+                if (!a.attributeMapper().allowMultiple() && !attrNames.add(a.attributeName())) {
+                    errors.add(new VerifyError("Multiple %s attributes in %s".formatted(a.attributeName(), toString(ae))));
+                }
+                verifyAttribute(ae, a, errors);
+            }
+        }
+        switch (cfe) {
+            case CompoundElement<?> comp -> {
+                for (var e : comp) verifyAttributes(e, errors);
+            }
+            case RecordAttribute ra -> {
+                for(var rc : ra.components()) verifyAttributes(rc, errors);
+            }
+            default -> {}
+        }
+    }
+
+    private void verifyAttribute(AttributedElement ae, Attribute<?> a, List<VerifyError> errors) {
+        int payLoad = ((BoundAttribute)a).payloadLen();
+        if (payLoad != switch (a) {
+            case AnnotationDefaultAttribute aa -> valueSize(aa.defaultValue());
+            case BootstrapMethodsAttribute bma -> {
+                int l = 2;
+                for (var bm : bma.bootstrapMethods()) {
+                    l += 4 + 2 * bm.arguments().size();
+                }
+                yield l;
+            }
+            case CharacterRangeTableAttribute cra -> 2 + 14 * cra.characterRangeTable().size();
+            case CodeAttribute ca -> {
+                MethodModel mm = (MethodModel)ae;
+                if (mm.flags().has(AccessFlag.NATIVE) || mm.flags().has(AccessFlag.ABSTRACT)) {
+                    errors.add(new VerifyError("Code attribute in native or abstract %s".formatted(toString(ae))));
+                }
+                if (ca.maxLocals() < Util.maxLocals(mm.flags().flagsMask(), mm.methodTypeSymbol())) {
+                    errors.add(new VerifyError("Arguments can't fit into locals in %s".formatted(toString(ae))));
+                }
+                int l = 12 + ca.codeLength() + 8 * ca.exceptionHandlers().size();
```

Comment:
nit (or not): it would be amazing if all these constants have a name. They are almost intuitive, I admit, but it would improve readability a lot.

Reply:
I think about the readability exactly opposite. When deciding between:
```
static final int CODE_ATTRIBUTE_HEADER_SIZE = 12;
static final int CODE_ATTRIBUTE_EXCEPTION_HANDLER_ENTRY_SIZE = 8;

...

int l = CODE_ATTRIBUTE_HEADER_SIZE + ca.codeLength() + CODE_ATTRIBUTE_EXCEPTION_HANDLER_ENTRY_SIZE * ca.exceptionHandlers().size();
```
or
```
int l = 12 + ca.codeLength() + 8 * ca.exceptionHandlers().size();
```
For this specific case ( implementation transcription of `classFileParser.cpp`) I prefer the shorter.